Add a new general-purpose codecov command

leandro-lucarella-sociomantic opened the pull request:
A new, general, `beaver codecov` command was added, which is not tied to *dlang* builds.

The new command is a wrapper for [codecov-bash](https://github.com/codecov/codecov-bash) (the same as `beaver dlang codecov` was), but assumes almost nothing about how to pass the reports.

Unlike `beaver dlang codecov`, the reports location must be passed explicitly to the script via the `BEAVER_CODECOV_REPORTS` environment variable. Glob patterns (like `*.lst`) and directories can be used (they will be copied recursively), but special characters are not escaped from the shell, so be careful if files have special characters, they must be properly escaped.

The `beaver dlang codecov` command is now based on the new `beaver codecov` command.

leandro-lucarella-sociomantic committed:
Add a new general-purpose codecov command

A new, general, `beaver codecov` command was added, which is not tied
to *dlang* builds.

The new command is a wrapper for
[codecov-bash](https://github.com/codecov/codecov-bash) (the same as
`beaver dlang codecov` was), but assumes almost nothing about how to
pass the reports.

Unlike `beaver dlang codecov`, the reports location must be passed
explicitly to the script via the `BEAVER_CODECOV_REPORTS` environment
variable. Glob patterns (like `*.lst`) and directories can be used (they
will be copied recursively), but special characters are not escaped from
the shell, so be careful if files have special characters, they must be
properly escaped.

The `beaver dlang codecov` command is now based on the new `beaver
codecov` command.

leandro-lucarella-sociomantic commented:
Untested, so testing is more than welcome.

mihails-strasuns-sociomantic commented on bin/dlang/codecov:
```diff
@@ -4,11 +4,14 @@
 
 # First check if there are any reports at all
-if test "$(echo .*.lst)" = ".*.lst"
+export BEAVER_CODECOV_REPORTS=".*.lst"
```
DMD generated lst files don't necessarily start with `.` - it only happens for files that are not compiled from current working dir as a root import path AFAIK. So, for example, integration test module reports will not have it.

leandro-lucarella-sociomantic replied:
Mmm, OK, but then this is a pre-existing bug :)

leandro-lucarella-sociomantic replied:
BTW, can you open an issue about this? It should be fixed in v0.3.x and this targets v0.4.x.

andreas-bok-sociomantic commented on README.md:
```diff
@@ -245,6 +245,39 @@ Travis already have a provider for deploying to bintray, but it is extremely
+
+The `[OPTIONS]` are forwarded directly to the
+[codecov-bash](https://github.com/codecov/codecov-bash), but this script will
+run in a confined docker instance where only the coverage reports are available,
```
So I wouldn't be able to just upload cached reports from a previous `travis` build stage?

leandro-lucarella-sociomantic replied:
Why not?

andreas-bok-sociomantic replied:
>  but this script will run in a confined docker instance where only the coverage reports are available,

As I understand from the description a new docker container will be started and the codecov bash script will be run inside that. OAIMS?

leandro-lucarella-sociomantic replied:
That is correct. And why do you think you would not be able to use cached reports?

andreas-bok-sociomantic commented on README.md:
```diff
@@ -245,6 +245,39 @@ Travis already have a provider for deploying to bintray, but it is extremely
+variables use `BEAVER_DOCKER_VARS` as usual.
+
+The reports location must be passed explicitly to the script via the
+`BEAVER_CODECOV_REPORTS` environment variable. Glob patterns (like `*.lst`) and
```
Just wondering why it is required to set the location of the report. This can already be specified passed to `codecov.bash` using `-f`. https://github.com/leandro-lucarella-sociomantic/beaver/blob/2c4d96cd009595b759eb713e80115856663eeea1/bin/codecov-bash#L89

And if the `[OPTIONS]` are parsed directly to the codecov.bash script the requirement of the env var **BEAVER_CODECOV_REPORTS** seems redundant.

leandro-lucarella-sociomantic replied:
> Just wondering why it is required to set the location of the report.

Because the reports need to be copied to the sandbox directory first.

andreas-bok-sociomantic replied:
> Because the reports need to be copied to the sandbox directory first.

Hmm oki. But for `libdrizzle-redux` where will the report be copied to? 
In https://github.com/sociomantic-tsunami/libdrizzle-redux/pull/271 the upload step isn't performed in a `docker` container but in the the native `travis` build environment, with the report file being uploaded directly from the cache-dir

leandro-lucarella-sociomantic replied:
> Hmm oki. But for libdrizzle-redux where will the report be copied to?

You don't care, that's taken care by beaver. Just tell beaver where to find the reports :)

leandro-lucarella-sociomantic replied:
> In sociomantic-tsunami/libdrizzle-redux#271 the upload step isn't performed in a docker container but in the the native travis build environment, with the report file being uploaded directly from the cache-dir

And when you switch to beaver, beaver will take care to build a sandbox for you, copy the reports to tell it to copy and run the `codecov` script in that sandbox. That's the whole point of using beaver :)

andreas-bok-sociomantic replied:
> And when you switch to beaver, beaver will take care to build a sandbox for you,

Oki, I will let beaver dig in  

![image](https://user-images.githubusercontent.com/13032446/35828821-328c59d2-0ac1-11e8-8d5f-79393d87d470.png)

leandro-lucarella-sociomantic replied:
😱

andreas-bok-sociomantic commented:
Super, I will test it with https://github.com/sociomantic-tsunami/libdrizzle-redux/pull/271

andreas-bok-sociomantic commented:
I tested with my coveralls PR. 
It seems like the `[OPTIONS]` are not passed correctly to the `codecov ` when using the example: 

    BEAVER_CODECOV_REPORTS="*.lst" beaver codecov [OPTIONS]` seems like the `[OPTIONS]

I don't run `gcov` so I pass `-X gcov` and `-Z` to exit with 1 on failure to `codecov`, thus running this command on `travis`: 

```
BEAVER_CODECOV_REPORTS="$CACHE_DIR/libdrizzle-redux-coverage.info" beaver codecov -X gcov -Z
beaver/bin/beaver: 20: exec: beaver/bin/codecov/-X: not found
```

leandro-lucarella-sociomantic commented:
Do you have a link to the travis log?

andreas-bok-sociomantic commented:
Here is the log: 

https://travis-ci.org/andreas-bok-sociomantic/libdrizzle-redux/jobs/337939131

leandro-lucarella-sociomantic committed:
beaver: Fail if the requested command doesn't exist

leandro-lucarella-sociomantic commented:
Right. 2 bugs found. Updated with fixes :D

leandro-lucarella-sociomantic commented:
BTW, fixes are:

1. Show an error if the requested beaver command doesn't exist (new commit)
2. Rename `bin/codecov` to `bin/beaver-codecov` (which is the correct name for `beaver` to find the command.

andreas-bok-sociomantic commented:
Does it work for the D tsunami projects? 
I tested last night with https://github.com/sociomantic-tsunami/libdrizzle-redux/pull/271 and it failed on travis. I test and post the errors.

andreas-bok-sociomantic commented:
AFAICS the path to the codecov script is still not resolved correctly. 
Also can't confirm whether arguments are passed correctly to `codecov` as I the error AFAICS happens before the `beaver-codecov` script  is called.  

I have linked to the errors on travis when testing with the latest version of this PR

travis log: https://travis-ci.org/andreas-bok-sociomantic/libdrizzle-redux/jobs/340871601


```
$ BEAVER_CODECOV_REPORTS="$CACHE_DIR/libdrizzle-redux-coverage.info" beaver codecov -X gcov -Z

cp: cannot stat ‘/home/travis/build/[secure]/libdrizzle-redux/bin/codecov-bash’: No such file or directory
The command "BEAVER_CODECOV_REPORTS="$CACHE_DIR/libdrizzle-redux-coverage.info" beaver codecov -X gcov -Z" exited with 1.
```


travis log: https://travis-ci.org/andreas-bok-sociomantic/libdrizzle-redux/jobs/340876737

```
$ BEAVER_CODECOV_REPORTS="$CACHE_DIR/libdrizzle-redux-coverage.info" beaver codecov

cp: cannot stat ‘/home/travis/build/[secure]/libdrizzle-redux/bin/codecov-bash’: No such file or directory
The command "BEAVER_CODECOV_REPORTS="$CACHE_DIR/libdrizzle-redux-coverage.info" beaver codecov" exited with 1.
```

leandro-lucarella-sociomantic commented:
Dammit, I'll have a look later today (hopefully).

leandro-lucarella-sociomantic commented:
OK, found the problem again, and this time I added a test too 😇

andreas-bok-sociomantic commented:
> OK, found the problem again, and this time I added a test too innocent

I will test

andreas-bok-sociomantic commented:
Now the problem is the dockerfile which beaver tries to use. The travis build is here : 

https://travis-ci.org/andreas-bok-sociomantic/libdrizzle-redux/jobs/341844421

Can I pass the `-f docker/Dockerfile.ubuntu.xenial` to `beaver codecov`?

leandro-lucarella-sociomantic commented:
> Now the problem is the dockerfile which beaver tries to use. The travis build is here :
> 
> https://travis-ci.org/andreas-bok-sociomantic/libdrizzle-redux/jobs/341844421
> 
> Can I pass the -f docker/Dockerfile.ubuntu.xenial to beaver codecov?

No, but it should use the latest one used (it assumes that `beaver install` or similar was used before.

leandro-lucarella-sociomantic commented:
Looking at the logs, I see you don't seem to be building the image, so I guess just adding `beaver install -f X` as you do with the jobs using docker before doing the `beaver codecov` should work.

leandro-lucarella-sociomantic commented:
Maybe we should mention this in the docs, but it would be good if you can try it out before.

andreas-bok-sociomantic commented:
> No, but it should use the latest one used (it assumes that beaver install or similar was used before.

Oki. 

> Looking at the logs, I see you don't seem to be building the image, so I guess just adding beaver install -f X as you do with the jobs using docker before doing the beaver codecov should work.

It works with `beaver install` but it requires ([as stated](https://github.com/sociomantic-tsunami/beaver#beaver-install)) adding a copy of the `docker/Dockerfile.ubuntu.xenial` file to the root renamed to `beaver.Dockerfile`. If not `beaver` reports that the `Dockerfile` is not found. 
Specifying the `Dockerfile` using `-f`isn't possible with `beaver install`. 

Success: https://travis-ci.org/andreas-bok-sociomantic/libdrizzle-redux/builds/343434546

Failed: https://travis-ci.org/andreas-bok-sociomantic/libdrizzle-redux/jobs/343446530

leandro-lucarella-sociomantic commented:
You can probably use the more "low-level" `beaver docker build`.

andreas-bok-sociomantic commented:
Tested with https://github.com/sociomantic-tsunami/libdrizzle-redux/pull/271, it works!! :)
So ready to go from libdrizzle-redux's side.

andreas-bok-sociomantic commented:
ping

leandro-lucarella-sociomantic commented:
Great, then merge is close, I added a note on the docker image used by the command and fixed a bit the docs specific to dlang.

```diff
diff --git a/README.md b/README.md
index 133f530..a945dde 100644
--- a/README.md
+++ b/README.md
@@ -262,6 +262,11 @@ run in a confined docker instance where only the coverage reports are available,
 plus the original source code as it was in Git (not the, possibly dirtly, build
 workspace).
 
+The docker images used to run the codecov script is determined by the standard
+`BEAVER_DOCKER_IMG` variable, and it is expected that this image was build
+before running `beaver codecov` (normally through the `beaver install` command
+or similar).
+
 Only `TRAVIS*` and `CODECOV_*` environment variables will be automantically
 passed to the docker container. If you want to pass any other environment
 variables use `BEAVER_DOCKER_VARS` as usual.
@@ -279,8 +279,8 @@ characters, they must be properly escaped.
 
 Some options are passed to codecov by default (at the moment `-n beaver -s
 reports`, where `reports` is the location of the sanitized sandbox where reports
-are copied; please check the `bin/dlang/codecov` script if you are interested in
-the details).
+are copied; please check the `bin/beaver-codecov` script if you are interested
+in the details).
 
 
 Building D1/2 projects
@@ -463,20 +463,15 @@ then use something like this in the `.travis.yml` file:
 after_success: beaver dlang codecov [OPTIONS]
 \```
 
-The `[OPTIONS]` are forwarded directly to the
-[codecov-bash](https://github.com/codecov/codecov-bash), but this script will
-run in a confined docker instance where only the coverage reports are available
-(this means no file list will be sent to codecov).
+This command is based on the global `beaver codecov` command (please read the
+documentation on this command for more details), but automatically selects which
+reports to send, so you don't need to pass the `BEAVER_CODECOV_REPORTS`
+environment variable. Also some extra variables specific to D programs are
+passed and used as flags (`DIST DMD DC F V` etc.). If you want to pass any other
+environment variables use `BEAVER_DOCKER_VARS` as usual.
 
-Only `TRAVIS*`, `CODECOV_*` and some dlang-related (`DIST DMD DC F V` etc.)
-environment variables will be automantically passed to the docker container. If
-you want to pass any other environment variables use `BEAVER_DOCKER_VARS` as
-usual.
-
-The following environment variables are reported automatically to codecov (via
-the `-e` option): `DIST`, `DMD`, `DC`, `F`. Some other options are passed to
-codecov by default, please check the `bin/dlang/codecov` script if you are
-interested in the details.
+Please check the `bin/dlang/codecov` script if you are interested in the
+details.
 
 
 ### Auto-convert and release tags for D2
```

leandro-lucarella-sociomantic commented:
I'll merge and do a RC. If there are any problems with the new docs we can fix them later before the final release as they are very minor.
